ansible_base/authentication/authenticator_plugins/ldap.py

```diff
@@ -104,7 +104,7 @@ def validator(value):
         self.validators.append(validator)
 
 
-def validate_ldap_filter(value: Any, with_user: bool = False) -> bool:
+def validate_ldap_filter(value: Any, with_user: bool = False) -> None:
```

Comment:
Okay, we're saying `value: Any` and then immediately raising `ValidationError` if it's not a string. If we are actually enforcing types, then theory says it should be safe to drop that check and just accept `value: str`.

But I'm not sure how practical this is, because it probably comes down to the JSON input, and there's no type-level validation at that point.

In other typed languages like Haskell the JSON deserializer would really return a [sum type](https://en.wikipedia.org/wiki/Tagged_union) that is something like `Either[Err, str]` and you'd handle both cases (`Err` being some error type returned by the JSON deserializer that says "hey, your JSON didn't make sense").

More succinctly: I find `value: Any` an the `type(value)` check really odd, but I believe fixing it would require a lot of work.

Reply:
> In other typed languages like Haskell the JSON deserializer would really return a [sum type](https://en.wikipedia.org/wiki/Tagged_union) that is something like Either[Err, str] and you'd handle both cases (Err being some error type returned by the JSON deserializer that says "hey, your JSON didn't make sense").

Sounds like what I was reading at https://stackoverflow.com/questions/44282268/python-type-hinting-with-exceptions

tl;dr - no error type hints in python ever.

> If we are actually enforcing types, then theory says it should be safe to drop that check and just accept `value: str`.

That would make this `typeguard` check fail because the test passes in a parameter in violation of the type annotation. This is not where you want it to fail, because you want the _method_ to raise its error, because the entire point of the test to to assert that the method raises that error.

That proves the point of the necessity of the decorator to turn off type enforcement.

But it also proves why this PR to add checks for correctness of type hints should avoid getting into the problem of adding type hints. If someone wanted to change the argument type hint, and thus turn off type checking in the test later, fine.